### PR TITLE
[Fluid-Sdem] qsvms dem coupled revision

### PR DESCRIPTION
**Pull request by jgonzalezusua**
**Description**
The aim of this PR is to revise that all terms of the element are correctly implemented using a manufactured solution to check it.

- Applications, Feature

**Changelog**

- Adding mass source in the RHS vector
- Modified manufactured solution to ensure it is derivable in space and in time
- Modified the projector to normalize the error
- Changed a line in _swimming_DEM_analysis.py_
- Modified the hdf5_script to let it write a new database file.

### CODE REVIEW

**Comment by jgonzalezusua** [comment]
Codacy warnings have been addressed except one super() that has to be written like in the code to make it functional.

**Comment by marcnunezc** [comment]
Hi!
CI failed do to an error unrealted to this PR.  Merge master to fix it! Sorry for the inconvenience.

**Comment by jgonzalezusua** [comment]
Ok, thanks, I will do it

**Review by rubenzorrilla — COMMENTED** [review]
No concerns from the `FluidDynamicsApplication` side.

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h [hunk 1]
```diff
@@ -78,10 +87,8 @@ double GetL2VectorProjection(ModelPart& r_model_part)
             error.push_back(error_x);
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: Could you not pre-reserve the memory for this vector insteadd of doing all those push_backs?

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h [hunk 2]
```diff
@@ -111,6 +118,15 @@ double GetL2ScalarProjection(ModelPart& r_model_part)
         NContainer = rGeom.ShapeFunctionsValues(GeometryData::GI_GAUSS_2);
         const SizeType& NumGauss = NContainer.size1();
 
+        element_area = 0.0;
+
+        for (unsigned int k = 0; k < NumNodes; ++k)
+        {
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: move this bracket to previous line

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h [hunk 3]
```diff
@@ -65,6 +65,15 @@ double GetL2VectorProjection(ModelPart& r_model_part)
...
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: move bracket to previous line

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h [hunk 4]
```diff
@@ -119,14 +135,14 @@ double GetL2ScalarProjection(ModelPart& r_model_part)
             for (unsigned int j = 0; j < NumNodes; ++j){
                 error += rGeom[j].FastGetSolutionStepValue(SCALAR_ERROR) * N[j];
             }
-            result += std::pow(error,2) * DetJ[gss] * IntegrationPoints[gss].Weight();
+            result += element_area * std::pow(error,2) * DetJ[gss] * IntegrationPoints[gss].Weight();
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: I think this is wrong... You are integrating the error. Why are you multiplying by the 'element area'? The Gauss Points already have that information.

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h [hunk 5]
```diff
@@ -111,6 +118,15 @@ double GetL2ScalarProjection(ModelPart& r_model_part)
         NContainer = rGeom.ShapeFunctionsValues(GeometryData::GI_GAUSS_2);
         const SizeType& NumGauss = NContainer.size1();
 
+        element_area = 0.0;
+
+        for (unsigned int k = 0; k < NumNodes; ++k)
+        {
+            element_area +=  ielem->GetGeometry()[k].FastGetSolutionStepValue(NODAL_AREA);
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: The total volume is too big here... The sum of the nodal areas is the total area... But if you ad one node at a time! Here you are going element by element and then running through the nodes, so some nodes are counted many times

**Review by GuillermoCasas — CHANGES_REQUESTED** [review]
There are issues in the implementation of the error norms. Please see comments

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h (hunk 4 above)
Comment (jgonzalezusua) [reply applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: Yes, it should be removed.

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h [hunk 6]
```diff
@@ -47,10 +47,23 @@ L2ErrorProjection()
...
+    const unsigned int n_nodes = r_model_part.Nodes().size();
+    for (unsigned int inode = 1; inode <= n_nodes; ++inode){
+
+        auto& r_vectorial_error = r_model_part.GetNode(inode).FastGetSolutionStepValue(VECTORIAL_ERROR);
+        r_vectorial_error = r_model_part.GetNode(inode).FastGetSolutionStepValue(VELOCITY) - r_model_part.GetNode(inode).FastGetSolutionStepValue(EXACT_VELOCITY);
+
+        auto& r_scalar_error = r_model_part.GetNode(inode).FastGetSolutionStepValue(SCALAR_ERROR);
+        r_scalar_error = r_model_part.GetNode(inode).FastGetSolutionStepValue(PRESSURE) - r_model_part.GetNode(inode).FastGetSolutionStepValue(EXACT_PRESSURE);
+    }
+}
```
Comment (philbucher) [inline applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: ouch this is completely wrong
- using `GetNode` is access by Id not by index
- using `GetNode` is a `find` each time, aka you search the same node 6 times per iteration!!!

I recommend using the parallel utilities (in pseudo code, please search the code for "block_for_each" to see usage examples):
~~~c++
block_for_each(r_model_part.Nodes(), [](Node<3>& rNode){
// here put your code
})
~~~

**Review by philbucher — CHANGES_REQUESTED** [review]
there is a severe bug in one of your functions

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h (hunk 6 above)
Comment (jgonzalezusua) [reply applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: Thanks @philbucher, I've already used this parallel utility in other scripts, I will apply this technique to the L2 projection as well.

**Review by philbucher — DISMISSED** [review]
now looks good to me

**Comment by jgonzalezusua** [comment]
> now looks good to me

I tried to add these `block_for_each` in the _transient_spatial_dependant_porosity_solution_body_force_process.cpp_ but I have had paralellization problems. Finally, I decided not to use them.

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h [hunk 7]
```diff
@@ -73,32 +86,27 @@ double GetL2VectorProjection(ModelPart& r_model_part)
...
-            error.push_back(error_x);
-            error.push_back(error_y);
-            if (dim == 3) error.push_back(error_z);
-            // for (unsigned int d = 0; d < dim; ++d){
-            //     scalar_product[d] = error[d];
-            // }
-            result += std::pow(SWIMMING_MODULUS_3(error),2) * DetJ[gss] * IntegrationPoints[gss].Weight();
-            error.clear();
+            modulus = std::sqrt(std::pow(error_x,2)+std::pow(error_y,2)+std::pow(error_z,2));
+            result += std::pow(modulus,2) * DetJ[gss] * IntegrationPoints[gss].Weight();
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: you are squaring a square root here... Please avoid it. Also look for a better name than 'interpolator'. Something that is descriptive of what it really is (something to do with the error).

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h [hunk 8]
```diff
@@ -110,6 +118,9 @@ double GetL2ScalarProjection(ModelPart& r_model_part)
         const GeometryType::IntegrationPointsArrayType& IntegrationPoints = ielem->GetGeometry().IntegrationPoints(GeometryData::GI_GAUSS_2);
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: use rGeom here. Also, I would use const auto

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h [hunk 9]
```diff
@@ -73,32 +86,27 @@ double GetL2VectorProjection(ModelPart& r_model_part)
...
         interpolator += result;
         result = 0.0;
     }
 
-    return std::sqrt(interpolator);
+    return std::sqrt(interpolator/total_area);
 }
 
 
 double GetL2ScalarProjection(ModelPart& r_model_part)
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: You are not 'projecting' anything here. You are calculating the L2-norm of the error. Change name accordingly

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h [hunk 10]
```diff
@@ -47,14 +47,25 @@ L2ErrorProjection()
...
+}
+
 double GetL2VectorProjection(ModelPart& r_model_part)
 {
     const unsigned int n_elements = r_model_part.Elements().size();
-    double interpolator = 0.0, result = 0.0, error_x = 0.0, error_y = 0.0, error_z = 0.0;
+    double modulus, total_area = 0.0, interpolator = 0.0, result = 0.0, error_x = 0.0, error_y = 0.0, error_z = 0.0;
     const unsigned int dim = r_model_part.GetProcessInfo()[DOMAIN_SIZE];
     Matrix NContainer;
     array_1d<double, 3> scalar_product;
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: This is not being used anymore

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h [hunk 11]
```diff
@@ -64,6 +75,8 @@ double GetL2VectorProjection(ModelPart& r_model_part)
         const GeometryType::IntegrationPointsArrayType& IntegrationPoints = ielem->GetGeometry().IntegrationPoints(GeometryData::GI_GAUSS_2);
         NContainer = rGeom.ShapeFunctionsValues(GeometryData::GI_GAUSS_2);
         const SizeType& NumGauss = NContainer.size1();
+        const double& Area = rGeom.Area();
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: 'Area' should be lowercase

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h [hunk 12]
```diff
@@ -64,6 +75,8 @@ double GetL2VectorProjection(ModelPart& r_model_part)
         const GeometryType::IntegrationPointsArrayType& IntegrationPoints = ielem->GetGeometry().IntegrationPoints(GeometryData::GI_GAUSS_2);
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: Use rGeom here

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h [hunk 13]
```diff
@@ -47,14 +47,25 @@ L2ErrorProjection()
 
 virtual ~L2ErrorProjection(){}
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: Here you are not really projecting, but merely calculating L2-norms. Please change name accordingly

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h (hunk 10 above)
Comment (jgonzalezusua) [reply applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: What do you mean by not being used?

Comment (GuillermoCasas) [reply applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: is the `scalar_product` variable used?

Comment (jgonzalezusua) [reply applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: Ok, you are right

**Comment by GuillermoCasas** [comment]
Thank you! Please address the remaining Codacy issue and I will approve.

**Review by GuillermoCasas — APPROVED** [review]

**Pull request merged**